### DISCUSSION TITLE
In your example you used closures was this just for simplicy or was there a julia specific reason

### DISCUSSION THREAD

**Discussion by ghost** [lbilli/Jib.jl#23] [Q&A] [answered]
Hi there,
I am a julia noob and am REALLY enjoying using Jib.jl as a template for learning.  I am trying to forget my python mindset and you approaches are wonderful for sending me off investigating WHY you took an approach and how I might use it in the future. Most of the time, with a LOT of help, I can figure out things BUT **the approach you took in the example you gave me ( EXCELLENT HELP by the way) used closures to build out the mdata Dict.  What was the design thinking there please OR was it just to provide a mwe.**

  Your code has been a wonderful learning tool once I dumped vscode and started using REPL and text editor.  I'm going to start using debugger.jl and Revise.jl when I really start my project.

**Comment by lbilli** [comment, marked as answer]
Normally data move in and out of functions via arguments and return values.
However, in an async context, callback functions (i.e. the members of `Wrapper`) are automatically called by code deep within the program that the user has no access to, therefore it's not possible to use the usual arguments and return values to pass data in and out. 
Alternative ways are required to allow callbacks to pass data in and out. 
Global variables, closures and channels are just examples of what can be used to this end.

**Comment by lbilli** [comment]
I'm not familiar with ZMQ, but if it uses regular Julia `Sockets` there shouldn't be any concern.
As packaged, the reader loop is concurrent but the program is still single threaded, so no issues about parts of the program stepping over each other.

Anyway, the [reader loop](https://github.com/lbilli/Jib.jl/blob/83366f43f435820c02e9363c7858a55006b1dad0/src/reader.jl#L65) is quite basic and can be easily customized.

**Reply by ghost** [reply]
thank you. I deleted the question as I thought it was out of the context of the OP. I am trying not to bother you with this stuff. Thank you for taking the time. REALLY helps. I am nervous to modify any of your code, it's working so well my end.
